File: python/mirage/core/github/scope.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import fnmatch

from mirage.cache.index import IndexEntry
from mirage.core.github.tree_entry import TreeEntry
from mirage.types import PathSpec


def scope_relative_key(path: PathSpec | str) -> str:
    """Strip the mount prefix from a path to get its repo-relative key.

    Args:
        path (PathSpec | str): Scope path, possibly mount-prefixed.

    Returns:
        str: Repo-relative key with a leading slash; ``/`` for the root.
    """
    prefix = path.prefix if isinstance(path, PathSpec) else ""
    key = path.original if isinstance(path, PathSpec) else str(path)
    if prefix and key.startswith(prefix):
        key = key[len(prefix):] or "/"
    return key


def is_repo_root(key: str) -> bool:
    """Return whether a repo-relative key points at the repository root.

    Args:
        key (str): Repo-relative key from :func:`scope_relative_key`.

    Returns:
        bool: True for ``""`` or ``"/"``.
    """
    return key in ("", "/")


def count_scope_files(entries: dict[str, IndexEntry], key: str) -> int:
    """Count indexed files under a repo-relative scope key.

    Args:
        entries (dict[str, IndexEntry]): Index entries keyed by repo-relative
            path with a leading slash.
        key (str): Repo-relative scope key from :func:`scope_relative_key`.

    Returns:
        int: Number of file entries at or below the scope.
    """
    if is_repo_root(key):
        return sum(1 for e in entries.values() if e.resource_type == "file")
    norm = "/" + key.strip("/")
    prefix = norm + "/"
    return sum(
        1 for p, e in entries.items()
        if e.resource_type == "file" and (p == norm or p.startswith(prefix)))


def should_use_search(
    is_regex: bool,
    recursive: bool,
    on_default_branch: bool,
) -> bool:
    """Whether grep/rg should narrow paths via GitHub code search.

    Search is appropriate for non-regex patterns running recursively on the
    default branch (code search only indexes the default branch).
    """
    return not is_regex and recursive and on_default_branch


async def estimate_scope(tree: dict[str, TreeEntry], directory: str,
                         pattern: str) -> tuple[int, int]:
    key = directory
    prefix = key + "/" if key else ""
    file_count = 0
    total_bytes = 0
    for p, entry in tree.items():
        if not p.startswith(prefix):
            continue
        remainder = p[len(prefix):]
        if entry.type == "blob" and fnmatch.fnmatch(remainder, pattern):
            file_count += 1
            total_bytes += entry.size or 0
    return file_count, total_bytes

File: typescript/packages/core/src/core/github/scope.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { fnmatch } from '../../util/fnmatch.ts'
import { stripSlash } from '../../util/slash.ts'
import type { PathSpec } from '../../types.ts'
import type { TreeEntry } from './tree_entry.ts'

export function scopeRelativeKey(path: PathSpec): string {
  const prefix = path.prefix
  let key = path.original
  if (prefix !== '' && key.startsWith(prefix)) {
    key = key.slice(prefix.length) || '/'
  }
  return key
}

export function isRepoRoot(key: string): boolean {
  return key === '' || key === '/'
}

export function countScopeFiles(tree: Record<string, TreeEntry>, key: string): number {
  // tree keys are repo-relative without a leading slash
  if (isRepoRoot(key)) {
    let count = 0
    for (const entry of Object.values(tree)) if (entry.type === 'blob') count += 1
    return count
  }
  const norm = stripSlash(key)
  const prefix = `${norm}/`
  let count = 0
  for (const [p, entry] of Object.entries(tree)) {
    if (entry.type !== 'blob') continue
    if (p === norm || p.startsWith(prefix)) count += 1
  }
  return count
}

export function shouldUseSearch(
  isRegex: boolean,
  recursive: boolean,
  onDefaultBranch: boolean,
): boolean {
  return !isRegex && recursive && onDefaultBranch
}

export function estimateScope(
  tree: Record<string, TreeEntry>,
  directory: string,
  pattern: string,
): { fileCount: number; totalBytes: number } {
  const key = directory
  const prefix = key !== '' ? `${key}/` : ''
  let fileCount = 0
  let totalBytes = 0
  for (const [p, entry] of Object.entries(tree)) {
    if (!p.startsWith(prefix)) continue
    const remainder = p.slice(prefix.length)
    if (entry.type === 'blob' && fnmatch(remainder, pattern)) {
      fileCount += 1
      totalBytes += entry.size ?? 0
    }
  }
  return { fileCount, totalBytes }
}
